### PR TITLE
[opentitanlib] Add alert_handler CRC32 magic value computation

### PR DESCRIPTION
**Pull request by jon-flatley**
As per PR #13468 the register values in alert_handler are read back as part of rom_pre_boot_check(), summed under a CRC32, and compared to a lifecycle specific value in OTP. This PR mimics the alert_handler register initialization to produce the same CRC32 value for programming into OTP.

### CODE REVIEW

File: sw/host/opentitanlib/src/otp/alert_handler_regs.rs [hunk 1]
```diff
@@ -0,0 +1,1215 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+/* automatically generated by rust-bindgen 0.60.1 */
```
Comment (alphan) [inline sw/host/opentitanlib/src/otp/alert_handler_regs.rs:5]: awesome! Can you regenerate this file because 621b69ea0c35c5d30373755f65fc4108fbceb0dc adds new alerts and I'm not sure if they are included here?

Also, can you
- [ ] Add the script that is used to generate this file so that we can regenerate it easily if needed, and
- [ ] Create an issue (high priority) to add a CI check for this file? I think we already have a similar check in the CI? @timothytrippel, right?

Thanks!

File: sw/host/opentitanlib/src/otp/lc_state.rs [hunk 2]
```diff
@@ -24,6 +24,15 @@ pub struct LcState {
     secded: LcSecded,
 }
 
+#[repr(u32)]
```
Comment (alphan) [inline sw/host/opentitanlib/src/otp/lc_state.rs:27]: nit: should we bindgen these as well?

If not, can you at least add a comment here and the corresponding header file?

File: sw/host/opentitanlib/src/otp/alert_handler.rs [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+        self.class[index] = reg;
+        self.class_regs[index].accum_thresh = config.accum_thresh;
+        self.class_regs[index].timeout_cyc = config.timeout_cyc;
+        self.class_regs[index].phase_cycs = config.phase_cycs;
+
+        Ok(())
+    }
+}
+
+trait CRC32Add {
```
Comment (alphan) [inline sw/host/opentitanlib/src/otp/alert_handler.rs:301]: nit: Should this be `Crc32Add`?

File: sw/host/opentitanlib/src/otp/alert_handler.rs [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+
+fn new_crc() -> Crc<u32> {
+    Crc::<u32>::new(&crc::CRC_32_ISO_HDLC)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    // A sanity test to make sure the correct CRC algorithm is being used.
```
Comment (alphan) [inline sw/host/opentitanlib/src/otp/alert_handler.rs:346]: nit : ~sanity~
nit: can you also add a reference to the crc32 unittest and add a reference to this file in the unit tests?

File: sw/host/opentitanlib/src/otp/alert_handler.rs [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+    loc_regwen: [u32; ALERT_HANDLER_LOC_ALERT_REGWEN_MULTIREG_COUNT as usize],
+    /// ALERT_HANDLER_LOC_ALERT_EN_SHADOWED
+    loc_en: [u32; ALERT_HANDLER_LOC_ALERT_EN_SHADOWED_MULTIREG_COUNT as usize],
+    /// ALERT_HANDLER_LOC_ALERT_CLASS_SHADOWED
+    loc_class: [u32; ALERT_HANDLER_LOC_ALERT_CLASS_SHADOWED_MULTIREG_COUNT as usize],
+    /// Alert handler class registers
+    class_regs: [AlertClassRegs; ALERT_HANDLER_PARAM_N_CLASSES as usize],
+}
+
+#[derive(TryFromPrimitive)]
```
Comment (alphan) [inline sw/host/opentitanlib/src/otp/alert_handler.rs:42]: We should use bindgen for these constants too.

File: sw/host/opentitanlib/src/otp/alert_handler.rs [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+    fn crc32_add(self, digest: &mut Digest<u32>) {
+        self.regwen.crc32_add(digest);
+        self.ctrl.crc32_add(digest);
+        self.accum_thresh.crc32_add(digest);
+        self.timeout_cyc.crc32_add(digest);
+        self.phase_cycs.crc32_add(digest);
+    }
+}
+
+impl CRC32Add for AlertRegs {
```
Comment (alphan) [inline sw/host/opentitanlib/src/otp/alert_handler.rs:327]: Did I miss it or the integration with the opentitantool will be covered in a separate PR? If so, please keep in mind that:
```
alert_config_crc32_for_lc_state = crc32 ^ lc_state ^ kErrorOk;
```

File: sw/host/opentitanlib/src/otp/alert_handler.rs (hunk 5 above)
Comment (alphan) [reply sw/host/opentitanlib/src/otp/alert_handler.rs:42]: That being said, we can address this in a follow-up, I think @cfrantz has a PR for this.

File: sw/host/opentitanlib/src/otp/alert_handler.rs [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+
+impl AlertRegs {
+    pub fn crc32(self) -> u32 {
+        let crc = new_crc();
+        let mut digest = crc.digest();
+        self.crc32_add(&mut digest);
+        digest.finalize()
+    }
+
+    pub fn new<T: OtpRead>(lc_state: LcStateVal, otp: &T) -> Result<Self> {
```
Comment (alphan) [inline sw/host/opentitanlib/src/otp/alert_handler.rs:134]: Thanks @jon-flatley! I know the mapping between the OTP values and alert_handler registers is far from trivial :)

TBH I didn't notice anything wrong but it's easy to miss a small error -- I think the best way to check this is to add e2e tests for several OTP configurations.

**Review by alphan — APPROVED** [review]
Thanks @jon-flatley! If you plan to handle the opentitantool update in a separate PR please see this [comment](https://github.com/lowRISC/opentitan/pull/14368#discussion_r960547975).

File: sw/host/opentitanlib/src/otp/alert_handler_regs.rs (hunk 1 above)
Comment (timothytrippel) [reply sw/host/opentitanlib/src/otp/alert_handler_regs.rs:5]: yeah thats right: https://cs.opensource.google/opentitan/opentitan/+/master:ci/scripts/check-generated.sh

File: sw/host/opentitanlib/src/otp/alert_handler.rs (hunk 7 above)
Comment (jon-flatley) [reply sw/host/opentitanlib/src/otp/alert_handler.rs]: Good suggestion, thanks @alphan. Please TAL at the additional test cases, one of which takes OTP values off the device to ensure it produces the same register values, and the other takes those register values and ensures that the correct CRC32 is produced.

File: sw/host/opentitanlib/src/otp/alert_handler.rs (hunk 5 above)
Comment (jon-flatley) [reply sw/host/opentitanlib/src/otp/alert_handler.rs:50]: I agree, looks like https://github.com/lowRISC/opentitan/pull/14860 covers this; will need to unify these

Comment (alphan) [reply sw/host/opentitanlib/src/otp/alert_handler.rs:50]: Just a quick note: #14860 was rolled-back in 28f79af3c. @timothytrippel do we track its return to the stage somewhere?

File: sw/host/opentitanlib/src/otp/alert_handler.rs [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+            LcStateVal::Test => return Ok(alert),
+        };
+
+        let class_enable = otp.read32("ROM_ALERT_CLASS_EN")?;
+        let class_escalate = otp.read32("ROM_ALERT_ESCALATION")?;
+
+        for i in 0..ALERT_HANDLER_ALERT_CLASS_SHADOWED_MULTIREG_COUNT as usize {
+            let value = otp.read32_offset(Some("ROM_ALERT_CLASSIFICATION"), i * 4)?;
+            let cls = AlertClass::try_from(
+                value.view_bits::<Lsb0>()[lc_shift..lc_shift + 8].load_le::<u8>(),
```
Comment (alphan) [inline sw/host/opentitanlib/src/otp/alert_handler.rs:165]: nit: Sorry for nitpicking but since we are accessing things in byte granularity, can we use sthg like `value.to_le_bytes()[byte_index]`?

File: sw/host/opentitanlib/src/otp/alert_handler.rs [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+                accum_thresh: 0x00000000,
+                timeout_cyc: 0x00000000,
+                phase_cycs: [0x00000000, 0x00000000, 0x00000000, 0x00000000],
+            },
+        ],
+    };
+
+    struct TestOtp {}
+
+    // OTP values that corrispond to the above `TEST_REG` values.
```
Comment (alphan) [inline sw/host/opentitanlib/src/otp/alert_handler.rs:457]:
```suggestion
    // OTP values that correspond to the above `TEST_REG` values.
```

File: sw/host/opentitanlib/src/otp/alert_handler.rs [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+
+fn new_crc() -> Crc<u32> {
+    Crc::<u32>::new(&crc::CRC_32_ISO_HDLC)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    // Register values dumped from device after alert_handler initialization.
```
Comment (alphan) [inline sw/host/opentitanlib/src/otp/alert_handler.rs:376]: 👍🏼

File: sw/host/opentitanlib/src/otp/alert_handler.rs [hunk 11]
```diff
@@ -0,0 +1,511 @@
...
+        self.loc_class.crc32_add(digest);
+        self.class_regs.crc32_add(digest);
+    }
+}
+
+fn new_crc() -> Crc<u32> {
+    Crc::<u32>::new(&crc::CRC_32_ISO_HDLC)
+}
+
+#[cfg(test)]
```
Comment (alphan) [inline sw/host/opentitanlib/src/otp/alert_handler.rs:372]: @milesdai @timothytrippel can you confirm that we are running rust unit tests in the CI?

**Review by alphan — APPROVED** [review]
Great work @jon-flatley! Just some nits + it looks like you need to run `bazel run //:buildifier_fix`. Let's get this merged asap so that we can integrate this with the `opentitantool`.

File: sw/host/opentitanlib/src/otp/alert_handler.rs (hunk 11 above)
Comment (milesdai) [reply sw/host/opentitanlib/src/otp/alert_handler.rs:358]: Yes, under `Earl Grey SW Build & Test`, we are basically running `bazel test --test_tag_filters=-broken,-cw310,-verilator,-dv /sw/...`, so as long as it's registered as a bazel test, it should be run.
